utils/network.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/vishvananda/netlink"
+)
+
+var slash32 = net.CIDRMask(32, 32)
+
+// DoNetworking performs the networking for the given config and IPAM result
+func DoNetworking(args *skel.CmdArgs, conf NetConf, result *types.Result) (string, string, error) {
+	hostVethName, contVethMac, err := setupContainerVeth(args.Netns, args.IfName, conf.MTU, result)
+	if err != nil {
+		return "", "", err
+	}
+
+	// Select the first 11 characters of the containerID for the host veth
+	newHostVethName := "cali" + args.ContainerID[:min(11, len(args.ContainerID))]
+	if err = setupHostVeth(hostVethName, newHostVethName); err != nil {
+		return "", "", err
+	}
+	return newHostVethName, contVethMac, nil
+}
+
+func setupContainerVeth(netns, ifName string, mtu int, res *types.Result) (string, string, error) {
+	var hostVethName, contVethMAC string
+	err := ns.WithNetNSPath(netns, func(hostNS ns.NetNS) error {
+		hostVeth, contVeth, err := ip.SetupVeth(ifName, mtu, hostNS)
+
+		if err != nil {
+			return err
+		}
+
+		// Handle IPv4
+		if res.IP4 != nil {
+			// Add a connected route to a dummy next hop so that a default route can be set
+			gw := net.IPv4(169, 254, 1, 1)
+			gwNet := &net.IPNet{IP: gw, Mask: slash32}
+			if err = netlink.RouteAdd(&netlink.Route{
+				LinkIndex: contVeth.Attrs().Index,
+				Scope:     netlink.SCOPE_LINK,
+				Dst:       gwNet}); err != nil {
+				return fmt.Errorf("failed to add route %v", err)
+			}
+
+			if err = ip.AddDefaultRoute(gw, contVeth); err != nil {
+				return fmt.Errorf("failed to add route %v", err)
+			}
+
+			if err = netlink.AddrAdd(contVeth, &netlink.Addr{IPNet: &res.IP4.IP}); err != nil {
+				return fmt.Errorf("failed to add IP addr to %q: %v", ifName, err)
+			}
+		}
+
+		// Handle IPv6
+		if res.IP6 != nil {
+			// No need to add a dummy next hop route as the device will already have an IPv6 LL address.
+			addresses, err := netlink.AddrList(contVeth, 6)
+			if err != nil {
+				return err
+			}
+
+			if len(addresses) < 1 {
+				return fmt.Errorf("Failed to get IPv6 addresses for container veth")
+			}
+
+			_, defNet, _ := net.ParseCIDR("::/0")
+			if err = ip.AddRoute(defNet, addresses[0].IP, contVeth); err != nil {
+				return fmt.Errorf("failed to add route %v", err)
+			}
+
+			if err = netlink.AddrAdd(contVeth, &netlink.Addr{IPNet: &res.IP6.IP}); err != nil {
+				return fmt.Errorf("failed to add IP addr to %q: %v", ifName, err)
+			}
+		}
+
+		hostVethName = hostVeth.Attrs().Name
+
+		contVeth, err = netlink.LinkByName(ifName)
+		if err != nil {
+			err = fmt.Errorf("failed to lookup %q: %v", ifName, err)
+			return err
+		}
+
+		contVethMAC = contVeth.Attrs().HardwareAddr.String()
+
+		return nil
+	})
+
+	return hostVethName, contVethMAC, err
+}
+
+func setupHostVeth(vethName, newVethName string) error {
+	// hostVeth moved namespaces and may have a new ifindex
```

Comment:
nit: Comment is not very helpful.  Should go into a bit more detail - "The host side veth was originally created within the container netns and was subsequently moved to the host netns.  The name will be the same, but the ifindex may have changed so we must re-query it."